shared/actions/signup.js

```diff
@@ -1,366 +1,304 @@
+  Saga.call(
+    RPCTypes.signupCheckInvitationCodeRpcPromise,
+    {invitationCode: action.payload.inviteCode},
+    Constants.waitingKey
+  )
+const checkInviteCodeSuccess = (_, action: SignupGen.CheckInviteCodePayload) =>
+  Saga.sequentially([
+    Saga.put(SignupGen.createCheckInviteCodeDone({inviteCode: action.payload.inviteCode})),
+    Saga.put(navigateTo([loginTab, 'signup', 'usernameAndEmail'])),
+  ])
+const checkInviteCodeError = () =>
+  Saga.put(SignupGen.createCheckInviteCodeDoneError({errorText: "Sorry, that's not a valid invite code."}))
+
+const requestAutoInvite = () =>
+  Saga.call(RPCTypes.signupGetInvitationCodeRpcPromise, undefined, Constants.waitingKey)
+const requestAutoInviteSuccess = (inviteCode: string) =>
+  Saga.put(SignupGen.createCheckInviteCode({inviteCode}))
+const requestAutoInviteError = () => Saga.put(navigateTo([loginTab, 'signup', 'inviteCode']))
+
+const requestInvite = (action: SignupGen.RequestInvitePayload) => {
+  const {email, name} = action.payload
+  const emailError = isValidEmail(email)
+  const nameError = isValidName(name)
+  if (emailError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError, name, nameError: ''}))
+  }
+  if (nameError) {
+    return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError: '', name, nameError}))
   }
+
+  return Saga.call(
+    RPCTypes.signupInviteRequestRpcPromise,
+    {email: email, fullname: name, notes: 'Requested through GUI app'},
+    Constants.waitingKey
+  )
 }
 
-function startRequestInvite() {
-  return (dispatch: Dispatch) => {
-    dispatch(SignupGen.createStartRequestInvite())
-    dispatch(nextPhase())
+const requestInviteSuccess = (
+  result: SignupGen.RequestInviteDonePayload | void,
+  action: SignupGen.RequestInvitePayload
+) => {
+  // rpc returns undefined, dispatches above on error return the type
+  if (result) {
+    return
   }
+  const {email, name} = action.payload
+  return Saga.sequentially([
+    Saga.put(SignupGen.createRequestInviteDone({email, name})),
+    Saga.put(navigateAppend(['requestInviteSuccess'], [loginTab, 'signup'])),
+  ])
 }
 
-function checkInviteCodeThenNextPhase(inviteCode: string) {
-  return (dispatch: Dispatch) => {
-    const p = new Promise((resolve, reject) => {
-      dispatch(SignupGen.createCheckInviteCode({inviteCode}))
+const requestInviteError = (err, action: SignupGen.RequestInvitePayload) => {
+  const {email, name} = action.payload
+  return Saga.put(SignupGen.createRequestInviteDoneError({email, emailError: err, name, nameError: ''}))
+}
 
-      RPCTypes.signupCheckInvitationCodeRpcPromise({
-        invitationCode: inviteCode,
-        waitingHandler: isWaiting => {
-          dispatch(SignupGen.createWaiting({waiting: isWaiting}))
-        },
+const checkUsernameEmail = (action: SignupGen.CheckUsernameEmailPayload) => {
+  const {email, username} = action.payload
+  const emailError = isValidEmail(email)
```

Comment:
Can you check both these errors and handle both in `checkUsernameEmailError`? Right now an email error takes precedence if both are invalid, but we could show both errors.
<img width="344" alt="image" src="https://user-images.githubusercontent.com/11968340/41920151-11f94334-792e-11e8-9abf-40a2745b9a22.png">